google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java

```diff
@@ -121,6 +129,324 @@
+      }
+    },
+    PARTITION_READ {
+      @Override
+      <ReqT, RespT> boolean isMethod(ReqT request, MethodDescriptor<ReqT, RespT> method) {
+        return method == SpannerGrpc.getPartitionReadMethod();
+      }
+    };
+
+    abstract <ReqT, RespT> boolean isMethod(ReqT request, MethodDescriptor<ReqT, RespT> method);
+
+    static <ReqT, RespT> SpannerMethod valueOf(ReqT request, MethodDescriptor<ReqT, RespT> method) {
+      for (SpannerMethod m : SpannerMethod.values()) {
+        if (m.isMethod(request, method)) {
+          return m;
+        }
+      }
+      return null;
+    }
+  }
+
+  /**
+   * Helper class to configure timeouts for specific Spanner RPCs. The {@link
+   * SpannerCallContextTimeoutConfigurator} must be set as a value on the {@link Context} using the
+   * {@link SpannerOptions#CALL_CONTEXT_CONFIGURATOR_KEY} key.
+   *
+   * <p>Example usage:
+   *
+   * <pre>{@code
+   * // Create a context with a ExecuteQuery timeout of 10 seconds.
+   * Context context =
+   *     Context.current()
+   *         .withValue(
+   *             SpannerOptions.CALL_CONTEXT_CONFIGURATOR_KEY,
+   *             SpannerCallContextTimeoutConfigurator.create()
+   *                 .withExecuteQueryTimeout(Duration.ofSeconds(10L)));
+   * context.run(
+   *     new Runnable() {
+   *       public void run() {
+   *         try (ResultSet rs =
+   *             client
+   *                 .singleUse()
+   *                 .executeQuery(
+   *                     Statement.of(
+   *                         "SELECT SingerId, FirstName, LastName FROM Singers ORDER BY LastName"))) {
+   *           while (rs.next()) {
+   *             System.out.printf("%d %s %s%n", rs.getLong(0), rs.getString(1), rs.getString(2));
+   *           }
+   *         } catch (SpannerException e) {
+   *           if (e.getErrorCode() == ErrorCode.DEADLINE_EXCEEDED) {
+   *             // Handle timeout.
+   *           }
+   *         }
+   *       }
+   *     });
+   * }</pre>
+   */
+  public static class SpannerCallContextTimeoutConfigurator implements CallContextConfigurator {
+    private Duration commitTimeout;
+    private Duration rollbackTimeout;
+
+    private Duration executeQueryTimeout;
+    private Duration executeUpdateTimeout;
+    private Duration batchUpdateTimeout;
+    private Duration readTimeout;
+
+    private Duration partitionQueryTimeout;
+    private Duration partitionReadTimeout;
+
+    public static SpannerCallContextTimeoutConfigurator create() {
+      return new SpannerCallContextTimeoutConfigurator();
+    }
+
+    private SpannerCallContextTimeoutConfigurator() {}
+
+    @Override
+    public <ReqT, RespT> ApiCallContext configure(
+        ApiCallContext context, ReqT request, MethodDescriptor<ReqT, RespT> method) {
+      SpannerMethod spannerMethod = SpannerMethod.valueOf(request, method);
+      if (spannerMethod == null) {
+        return null;
+      }
+      switch (SpannerMethod.valueOf(request, method)) {
```

Comment:
That's probably a good idea. We would still need some kind of lookup table for the timeout / method combination (i.e. keeping them in a hash map or something), but it might make it somewhat simpler. I'll do that in a separate PR, as it won't change the public interface of it.